### PR TITLE
[stdlib] Lemire’s Nearly Divisionless Random Integer Generation

### PR DESCRIPTION
**Pull request by palimondo**
Implementation of Daniel Lemire’s “Fast Random Integer Generation in Interval”
See https://arxiv.org/pdf/1805.10941.pdf

### CODE REVIEW

**Comment by palimondo** [comment]
I have attempted a Swift port of @lemire's algorithm:
````c
uint64_t nearlydivisionless ( uint64_t s, uint64_t (* random64 ) ( void ) ) {
  uint64_t x = random64 () ;
  __uint128_t m = ( __uint128_t ) x * ( __uint128_t ) s;
  uint64_t l = ( uint64_t ) m;
  if (l < s) {
    uint64_t t = -s % s;
    while (l < t) {
      x = random64 () ;
      m = ( __uint128_t ) x * ( __uint128_t ) s;
      l = ( uint64_t ) m;
    }
  }
  return m >> 64;
}
````

I'm not sure if I have correctly interpreted the negation of unsigned integer, I had to rely on https://stackoverflow.com/questions/8026694/c-unary-minus-operator-behavior-with-unsigned-operands
@stephentyrone can you please verify (since I've seen you answer that SO question...)?

````swift
@inlinable
public mutating func next<T: FixedWidthInteger & UnsignedInteger>(
  upperBound: T
) -> T {
  _precondition(upperBound != 0, "upperBound cannot be zero.")
  var random: T = next()
  var m = random.multipliedFullWidth(by: upperBound)
  if (m.low < upperBound) {
    let t = (T.max - upperBound + 1) % upperBound
    while (m.low < t) {
      random = next()
      m = random.multipliedFullWidth(by: upperBound)
    }
  }
  return m.high
}
````

Is there better way to express this in Swift than `(T.max - upperBound + 1)`

In my local test with `WyRand` on the worst case(?) bound `UInt64.max / 2 + 1` I'm seeing ~4x speed improvement in random number generation.

On a related note, is it even possible to change the implementation of `next(upperBound:)` this way given ABI stability constraints?

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by Azoy** [comment]
We are able to change the implementation of `next(upperBound:)` within ABI constraints. There's currently a notice on the documentation that states: 

> The algorithm used to create random values may change in a future version of Swift. If you’re passing a generator that results in the same sequence of integer values each time you run your program, that sequence may change when your program is compiled using a different version of Swift.

cc: @stephentyrone

**Comment by lemire** [comment]
Out of instinct, I would replace

```
 (T.max -upperBound + 1) % upperBound
```

by 

```
((~upperBound) &+ 1) % upperBound
```

or

```
(0 &- upperBound) % upperBound
```

depending on how clever you feel.

It might be useful to point to the reference as otherwise this function might look like mysterious magic to some: [Fast Random Integer Generation in an Interval](https://arxiv.org/abs/1805.10941), ACM Transactions on Modeling and Computer Simulation 29 (1), 2019

**Comment by Azoy** [comment]
I expect to see some performance wins with the floating point random API as well.

**Comment by palimondo** [comment]
Something's messed up. I need to investigate locally why has the benchmark crashed… but it's late over here, so I'll come back to this tomorrow.

**Comment by palimondo** [comment]
@lemire I _have_, of course, linked to your paper in the PR's description right at the top 😉

I liked the `((~upperBound) &+ 1) % upperBound` as it resembles your original C code, but looking at how's the [`~` implemented](https://github.com/apple/swift/blob/e9d4687e31a3ae8e90604d3b15bf8b241479c211/stdlib/public/core/Integers.swift#L2741) it's just 2 extra and unnecessary operations, so I'll go with `(0 &- upperBound) % upperBound` for a better unoptimized performance (I understand they all boil down to same instructions in `-O`).

**Comment by lemire** [comment]
@palimondo Sorry, I missed the first comment.

**Comment by palimondo** [comment]
Hmm.. reading through [Melissa O'Neil's post](http://www.pcg-random.org/posts/bounded-rands.html), I see there is still a "Bitmask with Rejection (Unbiased) — Apple's Method" that performed even slightly better…
@stephentyrone is there a way to express `mask >>= __builtin_clz(range|1);` in Swift?

**Comment by Azoy** [comment]
@palimondo `mask >>= (range | 1).leadingZeroBitCount`?

**Comment by lemire** [comment]
@palimondo 

I thought that Melissa's conclusion was that "The fastest (unbiased) method is Lemire's with an extra tweak from me [Melissa]."


Let us ping her : @imneme

**Comment by palimondo** [comment]
I'm sure you're right, I did not finish reading before posting here… 😇

**Comment by imneme** [comment]
@lemire, That _is_ what I said. Not sure I can add much beyond that…

Benchmarks are always something of a rabbit hole with more data the deeper you go but typically additional data reveals complexity rather than simple clarity. I really like @lemire's method in general. Frankly, I'd like it if there was a specific mod instruction itself was smart enough to realize when it has no work to do and short-circuit itself rather than needing any tricks. As it is (at least on x86), mod is really just a side effect of div and so it needs to go to the full amount of work, so my optimization is helpful.

(Possibly there is a general compiler optimization for mod, perhaps triggered by profiling!)

**Comment by lemire** [comment]
New blog post where I include ARM benchmarks: [Nearly Divisionless Random Integer Generation On Various Systems](https://lemire.me/blog/2019/06/06/nearly-divisionless-random-integer-generation-on-various-systems/).  I include benchmarks on an ARM box.

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
The optimized results seem quite nice. It even looks like the `-Osize` likes the new code a lot better than original. I dunno what weight to assign to the `-Onone` regressions…

I'm re-running benchmarks with @imneme's [modulo optimization](http://www.pcg-random.org/posts/bounded-rands.html#optimizing-modulo), to see how it fares in Swift.

**Comment by palimondo** [comment]
All results were identical with the exception of regressed `-O` `RandomShuffleLCG2`, but the difference comes to effectively just 2 μs (that test has `legacyFactor` of 16, meaning the real measured times were 44 μs and 46 μs), I'd say that was probably within the margin of error.

Looking at the generated assembly for the modulo optimization on Compiler Explorer, it looks like the [Swift version](https://swift.godbolt.org/z/B-aUUB) is not as tight [the C version](https://c.godbolt.org/z/uLVjUb) for some reason… Thoughts?

**Comment by lemire** [comment]
@palimondo One difference for sure is that Swift is careful regarding the case where the bound is zero (which leads to a division by zero) whereas the C version assumes the programmer checked.

File: stdlib/public/core/Random.swift [hunk 1]
```diff
@@ -94,15 +94,22 @@ extension RandomNumberGenerator {
...
-    var random: T = 0
-
-    repeat {
-      random = next()
-    } while random < range
-
-    return random % upperBound
+    var random: T = next()
+    var m = random.multipliedFullWidth(by: upperBound)
+    if (m.low < upperBound) {
```
Comment (xwu) [inline stdlib/public/core/Random.swift]: Nit:

```suggestion
    if m.low < upperBound {
```

...and elsewhere (to remove some C-isms).

**Comment by palimondo** [comment]
I realize that I have forgotten to use the masking subtraction assignment operator (`&-=`), and had just `t -= upperBound`, in the modulo optimization, but correcting that did not fix the generated code. I have tried bracketing the whole optimization in an explicit `upperBount !=0` if statement, but nothing I have tried made a difference… 

So I'm removing the modulo optimization until someone more clever than me figures out how to make it work in Swift.

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci test

**Comment by stephentyrone** [comment]
> So I'm removing the modulo optimization until someone more clever than me figures out how to make it work in Swift.

Please create two swift bugs if you haven't already:
- the missed optimization
- re-enabling the modulo optimization once we can

**Comment by stephentyrone** [comment]
Two things:

a. This method works well for small (sub-word) integer types, but is bad for larger types--we want to switch to rejection sampling (what Melissa calls "Apple's method", though it's so blindingly obvious that we can't have invented it) for types bigger than a word (this takes us from O(M(n)) down to O(n) complexity). That improvement doesn't need to be part of this PR, but is worth a bug report.

b. multipliedFullWidth currently fatalErrors for Int64 on 32b systems (https://github.com/apple/swift/blob/master/stdlib/public/core/IntegerTypes.swift.gyb#L1521). That's a bug, but it either needs to be fixed before this can go in, or you need to work around it here for now.

**Review by stephentyrone — CHANGES_REQUESTED** [review]
Two things:

a. This method works well for small (sub-word) integer types, but is bad for larger types--we want to switch to rejection sampling (what Melissa calls "Apple's method", though it's so blindingly obvious that Ian and I can't have invented it) for types bigger than a word (this takes us from O(M(n)) down to O(n) complexity). That improvement doesn't need to be part of this PR, but is worth a bug report.

b. multipliedFullWidth currently fatalErrors for Int64 on 32b systems (https://github.com/apple/swift/blob/master/stdlib/public/core/IntegerTypes.swift.gyb#L1521). That's a bug, but it either needs to be fixed before this can go in (I can do that sometime this week), or you need to work around it here for now, probably by conditionally using the old version on 32b systems.

**Comment by palimondo** [comment]
@stephentyrone I'm confused by what you mean in point a). Isn't the word size 64 bits on a 64bit system and Swift does [not yet support any bigger types](https://forums.swift.org/t/128-bit-operations-in-swift/21639 "Swift Forums: 128 bit operations in Swift?")?

I'm also unsure about the bug for missing optimization. If the reason for why O'Neill's optimization didn't work in Swift is the [division by zero check](https://github.com/apple/swift/blob/165fc4037c2d70592eae30332324ec77eaff076a/stdlib/public/core/IntegerTypes.swift.gyb#L1217) as [Lemire hypothesized](https://github.com/apple/swift/pull/25286#issuecomment-499860634), how would this be expressible without compromising the safety? Introduction of a new operator?

Anyway… I've filed [SR-10909 Missed optimization for modulo](https://bugs.swift.org/browse/SR-10909) and will add the architecture fork in next commit to keep the old implementation around for 32-bit platforms.

**Comment by stephentyrone** [comment]
> @stephentyrone I'm confused by what you mean in point a). Isn't the word size 64 bits on a 64bit system and Swift does [not yet support any bigger types](https://forums.swift.org/t/128-bit-operations-in-swift/21639)?

In the standard library today, yes. In third party code or in the standard library in the future, there may be arbitrary types conforming to FixedWidthInteger. Now that we have binary and module stability, future third-party types may even be back-deployed against *this* version of the standard library, so it is good not to let clear inefficiencies be introduced. I'm fine with this one in the short-term, because it's no worse than the current situation, but we should resolve it when we can.

> I'm also unsure about the bug for missing optimization. If the reason for why O'Neill's optimization didn't work in Swift is the [division by zero check](https://github.com/apple/swift/blob/165fc4037c2d70592eae30332324ec77eaff076a/stdlib/public/core/IntegerTypes.swift.gyb#L1217) as [Lemire hypothesized](https://github.com/apple/swift/pull/25286#issuecomment-499860634), how would this be expressible without compromising the safety? Introduction of a new operator?

Given `_precondition(upperBound != 0, "upperBound cannot be zero.")`, the compiler should be able to elide the check; it would have already been detected. I think this may be as simple as changing `_precondition` to something else--I don't think `_precondition` gives the needed hint to the compiler, and I recall there being a reason for it--but I'll need to dig in my notes.
 
> Anyway… I've filed [SR-10909 Missed optimization for modulo](https://bugs.swift.org/browse/SR-10909) and will add the architecture fork in next commit to keep the old implementation around for 32-bit platforms.

Thanks!

**Comment by palimondo** [comment]
@stephentyrone I hope you don't mind that I've created a reminder to add the [Support multipliedFullWidth on 32-bit systems (SR-10910)](https://bugs.swift.org/browse/SR-10910) for you… 😇

File: stdlib/public/core/Random.swift [hunk 2]
```diff
@@ -94,6 +94,7 @@ extension RandomNumberGenerator {
     upperBound: T
   ) -> T {
     _precondition(upperBound != 0, "upperBound cannot be zero.")
+#if arch(i386) || arch(arm)
```
Comment (stephentyrone) [inline stdlib/public/core/Random.swift]: Please add a TODO here and a swift bug to remove this when no longer needed.

Comment (palimondo) [reply stdlib/public/core/Random.swift]: [SR-10912](https://bugs.swift.org/browse/SR-10912) Use Lemire's Nearly Divisionless Random Integer Generation on 32-bit Systems

**Review by stephentyrone — APPROVED** [review]

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by stephentyrone** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
Do we have a way to benchmark this on 32-bit systems once SR-10910 gets resolved?

**Comment by stephentyrone** [comment]
I'm going to review this once more tomorrow morning for numerical correctness specifically and then merge it assuming it looks good. @airspeedswift and I are discussing cherry-picking for 5.1 as well.

**Comment by stephentyrone** [comment]
> Do we have a way to benchmark this on 32-bit systems once SR-10910 gets resolved?
> (Could it be necessary to keep the OpenBSD method for performance reasons on 32-bit platforms?)

No, because we'll want to have a rejection sampling path in the long run anyway for bigger-than-word types, which would also cover this case.

**Comment by stephentyrone** [comment]
Ended up having some time tonight. This is fine to merge.

**Comment by stephentyrone** [comment]
n.b. @lemire @imneme thinking about this, I believe that we can actually do somewhat better still by using straight-forward rejection sampling for large ranges and this algorithm for small ranges. This would be an extra branch, but it will be perfectly predicted in most uses (and could be hoisted or specialized for algorithms that produce values in bulk). I'll write up a demo sometime soon.

**Comment by palimondo** [comment]
@stephentyrone Didn't @imneme [try that here already](https://github.com/imneme/bounded-rands/blob/master/bounded64.cpp#L313-L336)?

**Comment by stephentyrone** [comment]
> @stephentyrone Didn't @imneme [try that here already](https://github.com/imneme/bounded-rands/blob/3d71f53c975b1e5b29f2f3b05a74e26dab9c3d84/bounded64.cpp#L313-L336)? That version was not in the charts of the [corresponding article](http://www.pcg-random.org/posts/bounded-rands.html#making-improvements) for some reason...

I would expect the perf tradeoff to be somewhat workload- (and hence benchmark-) dependent. It may be that it wasn't a substantial win for the use case that Melissa was testing, or there may be some issue we still need to understand, but for the top half of upperBound values, the actual selection looks the same with both approaches, and a comparison is certainly cheaper than a full-width multiply followed by a comparison. It likely doesn't matter for int64 due to the fact that the rejection branch is necessarily "unpredictable" with a good RNG, but for custom types where multiply can be much more costly than a branch mispredict (as this algorithm supports in Swift), it *will* matter.

**Comment by palimondo** [comment]
I meant it more like… "Is this the algorithm you had in mind? Maybe I could transcribe also that one…" 😜

**Comment by palimondo** [comment]
FWIW, I suspect I have misinterpreted the situation with modulo optimization. The Swift vs. C assembly comparison wasn't fair as I took a too narrow code snippet I suspected to be problematic and took out code that included important hints for the Swift compiler to perform optimization. Further, the original code was C++!

I re-did the Godbolt Compiler Explorer comparison again:
* [Swift|https://swift.godbolt.org/z/7KqwJk]
* [C++|https://cpp.godbolt.org/z/jGApEz]

I believe the generated instructions are largely equivalent.

Looking again at the original article by O'Neill, the modulo optimization provides a slight improvement on [Large-Shuffle Benchmark|http://www.pcg-random.org/posts/bounded-rands.html#large-shuffle-benchmark-results_1], but on [Small-Shuffle Benchmark|http://www.pcg-random.org/posts/bounded-rands.html#small-shuffle-benchmark-results_1] the version with modulo optimization performs the same as Lemire's original version "Debiased Int Mult (t-opt)" in chart. Our benchmarks in SBS most closely correspond to small shuffle, which would explain why I saw no improvement from modulo optimization on our benchmarks.

@stephentyrone I guess I should come up with a synthetic benchmark to mimic O'Neill's Large-Shuffle and we could re-introduce the modulo optimization together with your rejection sampling (when you come around to doing that).

**Comment by palimondo** [comment]
FWIW, I suspect I have misinterpreted the situation. The Swift vs. C assembly comparison wasn't fair as I took a too narrow code snippet I suspected to be problematic and took out code that included important hints for the Swift compiler to perform optimization. Further, the original code was C++!

I re-did the Godbolt Compiler Explorer comparison again:
* [Swift](https://swift.godbolt.org/z/7KqwJk)
* [C++](https://cpp.godbolt.org/z/jGApEz)

I believe the generated instructions are largely equivalent.

Looking again at the original article by @imneme, the modulo optimization provides visible improvement on [Large-Shuffle Benchmark](http://www.pcg-random.org/posts/bounded-rands.html#large-shuffle-benchmark-results_1), but on [Small-Shuffle Benchmark](http://www.pcg-random.org/posts/bounded-rands.html#small-shuffle-benchmark-results_1) the version with modulo optimization, `Debiased Int Mult (t-opt, m-opt)` in the chart, performs the same as Lemire's original version `Debiased Int Mult (t-opt)`. 

Large-Shuffle | Small-Shuffle
---|---
Large-Shuffle Results|Small-Shuffle Results

Our benchmarks in SBS most closely correspond to small shuffle, which would explain why I saw no improvement from modulo optimization on our benchmarks.

@stephentyrone I guess I should come up with a synthetic benchmark to mimic O'Neill's Large-Shuffle and we could re-introduce the modulo optimization together with your rejection sampling (when you get around to it).

**Comment by tempelmann** [comment]
Here's a little critique, and a suggestion at the end. I'm not actively involved in Swift development, not even using it, so this is the best place I could find to add comments.

I am worried that a mistake has been made when designing this class. Sadly it's too late now to correct it.

The mistake is that the doc missed out on explaining how to create a custom next() function, especially if one wants to create a PRNG with a reproducible output (using a seed value). The problem is that the original impl of upperBounds: was apparently using the lower bits of the UInt64, whereas the new one uses the higher bits. That leads to the ugly effect that when someone implemented a PRNG that only delivered a smaller range, e.g. in the lower 32 bits, this used to work before this change when he only pulled small ranges from it. But with this change, the PRNG now fails, always returning zero for randomized subranges.

The issue is shown here in a code example with sample output: https://gist.github.com/nicklockwood/cf1fde3eb24602883c17f94e63e490bc

This potential problem should have been known about - it's clearly documented in any theory of PRNGs. For instance, this [WP article](https://en.wikipedia.org/wiki/Linear_congruential_generator#Parameters_in_common_use) conveniently lists examples and shows that many of them do not return good randomized values in their lower bits.

As one comment above says, the API docs only warned that PRNGs might return _different_ values after an implementation change, but they did not consider to break it entirely - yet, this change did. Agreed?

To prevent this, there could have been an extra function that returns the range of significantly randomized values. This could be done by returning a divisor and max value, which in the gist example above would have been: divisor=1 and max= 233280, whereas for the system RNG it's possibly 2^32 and 2^32 (assuming only the upper 32 bit contain good random values). With that, any subrange function, like the upperBounds: one, would use this information to first reduce the value by this range before applying the upperBounds reduction.

In fact, adding such a function might still be valuable as it would (a) make future implementations of custom PRNGs via a custom next() function easier, and (b) would also help other code that uses the original next() result and create valid values from it.

**Comment by xwu** [comment]
> That leads to the ugly effect that when someone implemented a PRNG that only delivered a smaller range, e.g. in the lower 32 bits

> To prevent this, there could have been an extra function that returns the range of significantly randomized values.

The method `next() -> UInt64` is expected to produce 64 bits of usable randomness. If your source of randomness only gives you 32 usable bits at a time, then you need to get data from that source more than once before returning from `next()`. We can certainly document that requirement more clearly.

**Comment by tempelmann** [comment]
Yes, that would be better than the current affair. Still, people might struggle with coming up with 64 bit random gens, and the need for reproducible RNGs is not uncommon. So please provide some assistance with that if you can.

**Comment by lemire** [comment]
I agree with @xwu's answer and I think it is the correct one though I would point out that the documentation already states: "Returns a value from a uniform, independent distribution of binary data." If you have non-trivial cycles or you are offering a limited amount of randomness, it seems that you are clearly violating the "uniform + independent" ideal.

> This potential problem should have been known about - it's clearly documented in any theory of PRNGs. 

It is true that, historically, people (and standard libraries) have relied on random number generators that would produce, for example, 32-bit words with only 31 bits of randomness (or less). In fact, the history of pseudo-random number generators is filled with really bad generators.

However, today, we have many really fast pseudo-random generators that produce high quality random words: it is quite easy to have a really fast generator that will produce 64-bit words that will pass hard statistical tests (e.g., consider the PCG family). If you really need it, you can rather easily go for cryptographically strong generator. Sky is the limit! Twenty or thirty years ago, there were difficult to find and difficult to write, but it is now much easier. There is no good reason to default to whatever bad generator your standard C library offers if you care about quality.

> I'm thinking of random gens that keep giving something like 0, 1, 0, 1 etc. (cycle length 2) or 0, 0, 1, 1 etc. (cycle length 4) on the lowest bit often. 

I think that trying to accomodate obviously flawed pseudo-random number generators, like the ones you describe, is unnecessary because we now have (in 2020) a large collection of high-quality generators.

> To prevent this, there could have been an extra function that returns the range of significantly randomized values. 

It seems that you are asking for support for legacy generators.

> Still, people might struggle with coming up with 64 bit random gens, and the need for reproducible RNGs is not uncommon.

That's a fair and probably correct point.

Some of us have offered some implementations...
https://github.com/lemire/SwiftWyhash

I think that the standard library could be extended with a few options. I, for one, would be willing to provide a couple if the core team called for it.

**Comment by tempelmann** [comment]
Thanks for your clarifications. I must admit I'm a bit behind on this topic, having working on and with PRNGs in the 80s, mostly. Also, thanks for taking this seriously. I only came into this because I had been interested why someone's attempt at subclassing Swift's RNG did work in older Xcode versions but not in recent ones. I tried to figure this out by reading the docs and found that they didn't provide answers to this. So, I think adding some clarifying docs is the right way to resolve this for others in the future.

**Pull request merged**